pkg/container-hook/bpf/execruntime.bpf.c

```diff
@@ -0,0 +1,251 @@
+	struct file *file = ctx->file;
+	struct task_struct *task = ctx->task;
+	struct task_struct *current_task;
+
+	if (!file || !task)
+		return 0;
+
+	// Initialization should only be done once
+	if (tracer_group != 0 || tracer_tgid != 0)
+		return 0;
+
+	// The current file is not a fanotify file
+	if (fanotify_fops_addr == 0 || (__u64)(file->f_op) != fanotify_fops_addr)
+		return 0;
+
+	// Check the current process
+	current_task = (struct task_struct *) bpf_get_current_task();
+	current_task = BPF_CORE_READ(current_task, group_leader);
+	if (task != current_task)
+		return 0;
+
+	if (tracer_fanotify_fd == ctx->fd) {
+		// found it!
+		tracer_group = (u64) BPF_CORE_READ(file, private_data);
+		tracer_tgid = bpf_get_current_pid_tgid() >> 32;
+		return 0;
+	}
+
+	return 0;
+}
+
+SEC("kprobe/fsnotify_remove_first_event")
+int BPF_KPROBE(ig_fa_pick, struct fsnotify_group *group)
+{
+	u64 pid_tgid;
+	u32 tgid;
+
+	if (tracer_group != (u64)group)
+		return 0;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+	tgid = pid_tgid >> 32;
+
+	// Keep context for kretprobe/fsnotify_remove_first_event
+	bpf_map_update_elem(&ig_fa_pick_ctx, &pid_tgid, &group, 0);
+
+	return 0;
+}
+
+SEC("kretprobe/fsnotify_remove_first_event")
+int BPF_KRETPROBE(ig_fa_pick_x, struct fanotify_event *ret)
+{
+	struct fsnotify_group **groupp;
+	struct record *record;
+	u32 pid;
+	u64 pid_tgid;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+
+	groupp = bpf_map_lookup_elem(&ig_fa_pick_ctx, &pid_tgid);
+	if (!groupp)
+		return 0;
+
+	pid = BPF_CORE_READ(ret, pid, numbers[0].nr);
+
+	record = bpf_map_lookup_elem(&exec_args, &pid);
+	if (!record) {
+		// no record found but we need to push an empty record in the queue to
+		// ensure userspace understands that there is not record for this event
+		goto add_empty_record;
+	}
+
+	if (record->pid == tracer_tgid) {
+		// skip tracer: we don't want to trace ourselves
+		goto add_empty_record;
+	}
+
+	bpf_map_push_elem(&ig_fa_records, record, 0);
+
+	bpf_map_delete_elem(&ig_fa_pick_ctx, &pid_tgid);
+	return 0;
+
+add_empty_record:
+	bpf_map_push_elem(&ig_fa_records, &empty_record, 0);
+	bpf_map_delete_elem(&ig_fa_pick_ctx, &pid_tgid);
+	return 0;
+}
+
+static __always_inline int comm_skip(const u8 *comm)
+{
+	// "containerd-shim"
+	if (comm[0] == 'c' && comm[1] == 'o' && comm[2] == 'n' && comm[3] == 't' && comm[4] == 'a' && comm[5] == 'i' && comm[6] == 'n' && comm[7] == 'e' && comm[8] == 'r' && comm[9] == 'd' && comm[10] == '-' && comm[11] == 's' && comm[12] == 'h' && comm[13] == 'i' && comm[14] == 'm' && comm[15] == '\0')
+		return 0;
+	// "runc" without '\0'
+	if (comm[0] == 'r' && comm[1] == 'u' && comm[2] == 'n' && comm[3] == 'c')
+		return 0;
+	// "exe"
+	if (comm[0] == 'e' && comm[1] == 'x' && comm[2] == 'e' && comm[3] == '\0')
+		return 0;
+	// "conmon"
+	if (comm[0] == 'c' && comm[1] == 'o' && comm[2] == 'n' && comm[3] == 'm' && comm[4] == 'o' && comm[5] == 'n' && comm[6] == '\0')
+		return 0;
+	return 1;
+}
```

Comment:
btw, why was this code removed?

Reply:
In this function, comm is the caller, e.g. comm=containerd-shim when calling runc or comm=conmon when calling crun.

I needed to get the arguments of conmon to get the container name. By the time crun is called, the container name is no longer visible in the command line. But the caller of conmon is not always the same program. It could be podman or crio or something else. So filtering on constant strings as optimization didn't make sense anymore.